src/Features/Core/Portable/Completion/Providers/ImportCompletionProvider/AbstractTypeImportCompletionProvider.cs

```diff
@@ -78,79 +82,122 @@ private async Task AddCompletionItemsAsync(CompletionContext completionContext,
             var document = completionContext.Document;
             var project = document.Project;
             var workspace = project.Solution.Workspace;
-            var typeImportCompletionService = workspace.Services.GetService<ITypeImportCompletionService>();
+            var typeImportCompletionService = document.GetLanguageService<ITypeImportCompletionService>();
 
             // Find all namespaces in scope at current cursor location, 
             // which will be used to filter so the provider only returns out-of-scope types.
             var namespacesInScope = GetNamespacesInScope(document, syntaxContext, cancellationToken);
 
+            var tasksOfGettingTopLevelTypes = new ConcurrentSet<Task<Task>>();
+
             // Get completion items from current project. 
             var compilation = await project.GetCompilationAsync(cancellationToken).ConfigureAwait(false);
-            await typeImportCompletionService.GetTopLevelTypesAsync(project, HandlePublicAndInternalItem, cancellationToken)
-                .ConfigureAwait(false);
-
-            // Get declarations from directly referenced projects and PEs
+            tasksOfGettingTopLevelTypes.Add(typeImportCompletionService.GetTopLevelTypesAsync(project,
+                                                                        syntaxContext,
+                                                                        HandlePublicAndInternalItem,
+                                                                        delayItemHandling: true,
+                                                                        cancellationToken));
+
+            // Get declarations from directly referenced projects and PEs.
+            // This can be done concurrently because we don't add items to 
+            // CompletionContext until the loop is completed. 
             var referencedAssemblySymbols = compilation.GetReferencedAssemblySymbols();
-            foreach (var assembly in referencedAssemblySymbols)
-            {
-                cancellationToken.ThrowIfCancellationRequested();
 
-                // Skip reference with only non-global alias.
-                var metadataReference = compilation.GetMetadataReference(assembly);
-                if (metadataReference.Properties.Aliases.IsEmpty ||
-                    metadataReference.Properties.Aliases.Any(alias => alias == MetadataReferenceProperties.GlobalAlias))
+            // This can be parallelized because we don't add items to CompletionContext
+            // until all the tasks are completed.
+            var combinedTask = Task.WhenAll(
```

Comment:
`select`ing and using async funcs seems a little wonky to me.  It feels (to me at least) that it would be clearer to just have  `List<Task>` and to just do your foreach and add each task to it i.e. `tasks.Add(typeImportCompletionService.GetTopLevelTypesAsync(...)`.

You can even do the Task.Run stuff to ensure it's on a threadpool thread if you want.

Reply:
Fixed